src/OpenTelemetry/Metrics/Exemplar/FixedSizeExemplarReservoir.cs

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Metrics;
+
+internal abstract class FixedSizeExemplarReservoir : ExemplarReservoir
+{
+    private readonly Exemplar[] runningExemplars;
+    private readonly Exemplar[] snapshotExemplars;
+
+    protected FixedSizeExemplarReservoir(int capacity)
+    {
+        Guard.ThrowIfOutOfRange(capacity, min: 1);
+
+        this.runningExemplars = new Exemplar[capacity];
+        this.snapshotExemplars = new Exemplar[capacity];
+        this.Capacity = capacity;
+    }
+
+    internal int Capacity { get; }
+
+    /// <summary>
+    /// Collects all the exemplars accumulated by the Reservoir.
+    /// </summary>
+    /// <returns><see cref="ReadOnlyExemplarCollection"/>.</returns>
+    public sealed override ReadOnlyExemplarCollection Collect()
+    {
+        if (this.ResetOnCollect)
+        {
+            for (int i = 0; i < this.runningExemplars.Length; i++)
+            {
+                ref var running = ref this.runningExemplars[i];
+                if (running.IsUpdated())
+                {
+                    running.Copy(ref this.snapshotExemplars[i]);
+                    running.Reset();
+                }
+                else
+                {
+                    this.snapshotExemplars[i].Reset();
+                }
+            }
+
+            this.OnReset();
+        }
+        else
+        {
+            for (int i = 0; i < this.runningExemplars.Length; i++)
+            {
+                ref var running = ref this.runningExemplars[i];
+                if (running.IsUpdated())
+                {
+                    running.Copy(ref this.snapshotExemplars[i]);
+                }
+                else
+                {
+                    this.snapshotExemplars[i].Reset();
+                }
+            }
+        }
+
+        this.OnCollected();
+
+        return new(this.snapshotExemplars);
+    }
+
+    internal sealed override void Initialize(AggregatorStore aggregatorStore)
+    {
+        var viewDefinedTagKeys = aggregatorStore.TagKeysInteresting;
+
+        for (int i = 0; i < this.runningExemplars.Length; i++)
+        {
+            this.runningExemplars[i].ViewDefinedTagKeys = viewDefinedTagKeys;
+            this.snapshotExemplars[i].ViewDefinedTagKeys = viewDefinedTagKeys;
+        }
+
+        base.Initialize(aggregatorStore);
+    }
+
+    protected virtual void OnCollected()
+    {
+    }
+
+    protected virtual void OnReset()
```

Comment:
Do we need `OnReset` as well in addition to `OnCollected`? There's no implementation of it as well.

Reply:
I'll remove it but it may be needed in the future. I have this spec issue open: https://github.com/open-telemetry/opentelemetry-specification/issues/3891

The spec says for SimpleFixedSizeExemplarReservoir:

> Any stateful portion of sampling computation SHOULD be reset every collection cycle. For the above example, that would mean that the num_measurements_seen count is reset every time the reservoir is collected.

> If no size configuration is provided, the default size MAY be the number of possible concurrent threads (e.g. numer of CPUs) to help reduce contention

This seems to lead (IMO) to some very odd behavior for cumulative temporality.

Lets say we have 10 CPUs so we have `SimpleFixedSizeExemplarReservoir` with storage for 10 exemplars.

* Delta

   Exemplar A, B, C offered.
   Collect runs. A, B, C exported.
   Exemplar D offered.
   Collect runs. D exported.

   👍 looks good

* Cumulative

   Exemplar A, B, C offered.
   Collect runs. A, B, C exported.
   Exemplar D offered.
   Collect runs. D, B, C exported.

   Because of the reset we always begin filling at the front which means some older stuff may linger around in the reservoir. This seems wrong to me. I think a more sensible design would be to always do what we do for delta, or just not reset for cumulative so we always weight things the same 🤷

I was using `OnReset` to change this behavior on #5364 but I didn't bring it into this PR.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/5386/commits/5cd161738019dccdeb651df5124ae507d6076e9a